backend/remote-state/gcs/client.go

```diff
@@ -77,6 +83,58 @@ func (c *remoteClient) Delete() error {
 	return nil
 }
 
+// createLockFile writes to a lock file, ensuring file creation. Returns the
+// generation number.
+func (c *remoteClient) createLockFile(lockFile *storage.ObjectHandle, infoJson []byte) (int64, error) {
+	w := lockFile.If(storage.Conditions{DoesNotExist: true}).NewWriter(c.storageContext)
+	if _, err := w.Write(infoJson); err != nil {
+		return 0, err
+	}
+	if err := w.Close(); err != nil {
+		return 0, err
+	}
+
+	return w.Attrs().Generation, nil
+}
+
+// unlockIfStale force-unlocks the lock file if it is stale. Returns true if a
+// stale lock was removed (and therefore a retry makes sense), otherwise false.
+func (c *remoteClient) unlockIfStale(lockFile *storage.ObjectHandle) bool {
+	if attrs, err := lockFile.Attrs(c.storageContext); err == nil {
+		age := time.Now().Sub(attrs.Updated)
+		if age > time.Duration(int64(minMissesUntilStale)*tickInterval.Nanoseconds()) {
+			if err := c.Unlock(strconv.FormatInt(attrs.Generation, 10)); err != nil {
+				log.Printf("[WARN] failed to release stale lock: %s", err)
+			} else {
+				return true
+			}
+		}
+	}
+
+	return false
+}
+
+// updateLockFile periodically updates the "updated" timestamp of the lock
+// file until the lock is released in Unlock().
+func (c *remoteClient) updateLockFile(lockFile *storage.ObjectHandle, generation int64) {
+	for _ = range c.ticker.C {
+		if attrs, err := lockFile.Attrs(c.storageContext); err == nil {
+			if attrs.Generation != generation {
```

Comment:
I think this should probably signal some sort of error, as there state lock was changed unexpectedly. We don't have a very good way of surfacing this in terraform, but other clients report it as an error during Unlock, so the user can at least manually verify the situation.

Reply:
Good point. I think this is already covered in a sense that Unlock will also fail when attempting to remove the lock file using the wrong generation number:

```go
func (c *remoteClient) Unlock(id string) error {
...
        if err := c.lockFile().If(storage.Conditions{GenerationMatch: gen}).Delete(c.storageContext); err != nil {
                return c.lockError(err)
        }

        return nil
}
```

In any case, added some log output.